python/ray/train/v2/_internal/state/state_actor.py

```diff
@@ -28,13 +49,75 @@ def __init__(self):
             self._is_train_run_export_api_enabled,
             self._is_train_run_attempt_export_api_enabled,
         ) = self._init_export_logger()
+        # TODO: consider row level locking.
+        self._runs_lock = threading.Lock()
+        self._run_attempts_lock = threading.Lock()
+
+        if ray_constants.env_bool(
+            ENABLE_STATE_ACTOR_POLLING_ENV_VAR,
+            DEFAULT_ENABLE_STATE_ACTOR_POLLING,
+        ):
+            self._poll_interval_s = float(
+                os.getenv(
+                    STATE_ACTOR_POLL_INTERVAL_S_ENV_VAR,
+                    DEFAULT_STATE_ACTOR_POLL_INTERVAL_S,
+                )
+            )
+            self._latest_poll_time = float("-inf")
+            self._start_controller_polling_thread()
+
+    def _abort_dead_controller_live_runs(self) -> None:
+        # Get runs to abort.
+        with self._runs_lock:
+            non_terminal_runs = [
```

Comment:
We release the lock, but then still modify / use it later in the loop

Reply:
Changed the implementation as per our offline discussion.